### DISCUSSION TITLE
How do I add line breaks for Real Live?

### DISCUSSION THREAD

**Discussion by Linden10** [satan53x/SExtractor#114] [Q&A] [answered]
Ello! I'm currently doing a translation of this game called "種をつける男〜めざせ全員受胎〜" released by CONCEPT back in 2005.
Also known as `"Tane o Tsukeru Otoko ~Mezase Zen'in Jutai~"` aka The Seed-Planting Man...if you know the meaning, yeah...
The cover of the game alone on vndb explains it's meaning well (A girl in a potted plant after...yup!)
VNDB Link: https://vndb.org/v5484

This game runs on the RealLive engine and is easy to extract the SEEN.txt for translation and thanks to SExtractor (Newest Version as of typing this), I'm able to convert the Seen####.txt into xlsx files for easy editing and it works in-game as seen in the screenshot below!

![TANE Translation WIP - Alice Girl - Cute and Funny](https://github.com/user-attachments/assets/117966a9-14c7-458f-8190-f7e8c1389113)
> Very cute and funny loli, Alice heh. Yes she's a heroine.

Oh and I used the arc_conv解包 in tools to extract the SEEN.txt file, not requiring decrypting as the game's SEEN.txt is fully unencrypted and easy to work with! ^^

Enabling "Generate VNT's JIS File" somehow also fixes issues with strange errors in the text too!

Now how am I translating this game? I'm using Translator++ with Google GeminiAI addon to translate the text entirely as recently, google's newest model as of last month or so has improved Japanese Translation a lot in terms of accuracy and context compared their previous models!

There's a separate issue with extracting the seen.txt files made by Translator++ with SExtractor though however so I just copy-and-paste the translated lines into the xlsx generated by SExtractor to make it work!

With that said, I like to ask for help on something I'm having trouble figuring out...
How do I add line breaks?

I tried various methods to no avail, from adding \r into the text to \n as well but the game doesn't line break but just display them like this:
![TANE Translation WIP - Line Breaks not working](https://github.com/user-attachments/assets/76866410-ade2-402a-bb98-36c64d42f67b)
> \n doesn't turn into a line break but is instead included in the text...

Also for some reason, commas ","and exclamation points "!" don't render at all in the game's text so I'm using full-width "、" and full-width "！"  in place which work well. If there's a way to make normal half-width commas and exclamation points appear in game, do tell!

Anyways that's all I like to ask and thank you for making such an amazing tool!
Please help and see ya!

**Comment by Linden10** [comment]
Oh and here are the settings I use for extracting and injecting the translated xlsx files into the seen####.txt files:
![SExtractor Settings 1](https://github.com/user-attachments/assets/cfcc3674-285c-4139-bf4c-b054618f48ee)

![SExtractor Settings 2](https://github.com/user-attachments/assets/273875d4-1f46-47cc-88f1-e2db1a77978d)

I appreciate that the settings have been partially translated to English by the way!

Anyways If there's anything that can be turned on to make line breaks work, do tell!
I'm not an expert at using this tool knowing that the instructions are in Chinese (however I learned to understand somewhat how it works).

OH! I already read this and saw your comment on adding line breaks but still had issues figuring out how to, hope you can still help! ^^
https://github.com/satan53x/SExtractor/issues/1#issuecomment-1694177186

**Comment by satan53x** [comment, marked as answer]
The `sample=<引擎选RealLive时不要用这个正则，这是BIN暴力提取用的，RealLive引擎提取不到时才用这个>` in regex means:
`<Don't use this regular expression when RealLive is selected as the engine. This is used for brute force extraction of BIN. Use this when the RealLive engine cannot extract it>`.
Prioritize using `_None`, which is the default regular expression of RealLive. 
It will turn line breaks into `<n>`, so you can add extra line breaks.
(If the game does not have secondary encryption, you need to remove the `decrypt=6`)

**Reply by Linden10** [reply]
Ah really! Thank you thank you, I'll test it out then!

Well I did test it out originally but when it ran, it took up almost all of my computer memory and nearly crash my computer before I forced closed it haha but I guess it was trying to decrypt an already unencrypted script!

Hopefully removing decrypt=6 means it'll run without potentially crashing my computer! 🤣

**Reply by Linden10** [reply]
Ah it didn't crash! Thank goodness!
Gonna test line breaks now!

**Comment by Linden10** [comment]
IT WORKED!!! Thank you thank you, using <n> in the text like so:
`　After enduring a battery of tests at the university<n>hospital's urology department, I returned home and collapsed<n>onto my bed, unable to even touch the convenience store<n>bento I'd bought on the way.`

converted them into linebreaks with SExtractor as seen below!
![TANE Translation WIP - Line Breaks are working](https://github.com/user-attachments/assets/bc26fa72-9dff-46d8-b18e-a51eb2b181ee)

This is fantastic news! Thanks again for helping @satan53x ^_^
Say before I close this, is there a way to make half-width commas and exclamation points work in RealLive?

That's all and see ya!

**Comment by satan53x** [comment]
In RealLive, some half-width symbol are used to control, so only can use the full-width such as `，` and `！`.

BTW, you can choose the `Encoding applies to BIN` to import text with `cp932`. (But not cache)
Or change the code to `self.NewEncodeName = 'cp932'` in `src/var_extract.py` if you usually only translate into English.

**Reply by Linden10** [reply]
Ah now I see, thanks for explaining! I guess I'll convert all the , and ! into their full-width forms!

Hmm I know that "Encoding applies to BIN" means it applies the encoding to binary files, but I guess you mean that the .txt files which are binary can also be imported with cp932 (Shift-JIS) encoding?

So...does that mean it'll automatically import the translations into that encoding, fixing the half-width/comma or...something else?
I'm a bit confused haha...

**Reply by satan53x** [reply]
The default import encoding is gbk in BIN, so importing full-width characters will cause game display errors. 
When you check `Generate VNT's JIS File`, you are actually changing the import encoding to shift-jis (cp932).

If you need to automatically convert some characters to full-width, you can edit the `text_conf.json` in SE root directory or copy the json into `ctrl` folder. And edit it as:
```
  "trans_replace": {
    "cp932": {
      "!": "！",
      ",": "，"
    }
  },
```
And check the `Translation Replacement` when import.

**Reply by Linden10** [reply]
AH NOW I UNDERSTAND! That answers the mysterious question of why "Generate VNT's JIS File" made the translations display correctly! It was because it when it was off, since the .txt files are binary they were being imported in the GBK encoding!

That's why I need to enable that setting to also include binary for Shift-JIS encoding heh, guess I can disable VNT since I don't need it!
THANK YOU!

Also thanks for the trans replace setting too, that'll make translating the game super easy alright! ^_^

Say say, do you think it's possible to also automatically replace line breaks like in this example:
```
　After enduring a battery of tests at the university
hospital's urology department, I returned home and collapsed
onto my bed, unable to even touch the convenience store
bento I'd bought on the way.
```

with `<n>` too? That'll be super awesome if possible!
It'll cut the time with translating the game too where I don't have to manually add `<n>` to each line! ^^

**Reply by satan53x** [reply]
Maybe you can try edit the text_conf: (Not tested)
```
  "orig_replace": {
    "<n>": "\n"
  }
```
No need to add trans_replace because the `orig_replace` is also responsible for restoring translations when import.

**Reply by satan53x** [reply]
The reason why `\n` was not used before is that it may conflict with the `Paragraph Separator` in the setting. 
You may need to change the `Paragraph Separator` to any string that will never appear in translation text, such as `abcde`.

**Reply by Linden10** [reply]
Nice and thanks! Hmmm...so I don't need to add trans_replace since orig_replace restores the translations when importing...

So I guess I just add
```
"cp932": {
      "!": "！",
      ",": "，"
```

to orig_replace and it'll do the same thing or no?
I'm a bit confused...

Ah I'll test orig_replace to see if it works! Thank you again!
AH I see why \n didn't work, conflicting with Paragraph Separator...
I'll leave it be if the orig_replace can automatically detect and replace the line breaks in the trans xlsx file with <n> unless...it's required to change Paragraph Separator to something else.

**Reply by satan53x** [reply]
`orig_replace` is responsible for both original text replacement and translation restoration. 
It means replace `<n>` to `\n` when extract and restore `\n` to `<n>` when import.

**Reply by satan53x** [reply]
> AH I see why \n didn't work, conflicting with Paragraph Separator...

Not really, the RealLive doesn't support \n. (\n is a conctrol symbol too)
In fact, the linebreak is converted into a byte array structure internally, so it is different from other engines that directly output it as \n.

**Reply by Linden10** [reply]
> `orig_replace` is responsible for both original text replacement and translation restoration. It means replace `<n>` to `\n` when extract and restore `\n` to `<n>` when import.

Okay...now I get it! So it means that if in the orig files that there's /n, SExtractor will extract and replace /n with <n>? If so I got it and so when it's imports the translation, it'll convert the <n> back into /n which is nice!

So...about that, the original scripts don't have line breaks haha.
I'm manually adding the line breaks myself in the translation files, so basically I might need to use "trans_replace" in this case as I like to replace the manually added linebreaks I added into the text into <n> which would make everything easier if possible.

Can I do that with trans_replace? Thanks for helping!

**Reply by satan53x** [reply]
If there's no linebreak in original scripts it means `orig_replace` only restore `\n` to `<n>` when import. 
It has the same effect as trans_replace `"\n": "<n>"`.

**Reply by Linden10** [reply]
I see! Will I'll test it out now and see if it works, thank you!

**Reply by Linden10** [reply]
IT WORKS! Orig_Replace actually worked with replacing the line breaks I made with `<n>`, yay~
Thank you!

Now I'm going to test using Trans Replace as well as I just tested Orig_Replace with the half-width to full-width code you shared earlier and it didn't work because it couldn't find the , in the orig files.

Yup! ^^

**Reply by Linden10** [reply]
Ah I got the same error as before! Since it's trying to detect if the half-width comma is in the orig files, when it should be focusing only on the translated xlsx files.
`json.decoder.JSONDecodeError: Expecting ',' delimiter: line 10 column 5 (char 133)`

This happens in both orig_replace and trans_replace when using this:
```
    "cp932": {
      "!": "！",
      ",": "，"
    }
```

Should I add "IgnoreDecodeError" or something?

**Reply by satan53x** [reply]
No, this is just an error of editing json. 
In json, commas must be added except for the last element.
Check your text_conf.json, there may be red error prompts in text editor.

**Reply by Linden10** [reply]
> No, this is just an error of editing json. In json, commas must be added except for the last element. Check your text_conf.json, there may be red error prompts in text editor.

I just realized what you mean...I forgot to add a comma knowing this is a json file...
Since they're two nested options, a comma is needed! My bad haha.

It's now working, thank you for pointing out my mistake! Now to test if it replaced them all well!

**Comment by satan53x** [comment]
`limit_maxlen_json_gt.py`: https://github.com/satan53x/SExtractor/issues/106#issuecomment-2395010748
You can use this script to process English translations, which will automatically re-split the text according to the maximum number of words per line.

**Reply by Linden10** [reply]
Ah...I feel I should mention but I realized that with all this infomation you shared, I don't need to worry about maxlen as apparently RealLive is not having any issue at all with lines that are longer then the text box supports, automatically allowing for the line to appear in the next box like so below:
![TANE Translation WIP - RealLive automatically carries text into next window](https://github.com/user-attachments/assets/4d2842c6-61b6-4a78-a480-4258fde93260)

Still thanks for the suggestion! I think this will be useful for another Visual Novel that might work with Max len heh.

**Comment by Linden10** [comment]
yay it worked! Now I'll be able to translate in peace! Thank you very much for everything @satan53x
You're the best!

This won't be the last time I'll ask for your help as they're some games I'll also ask for help heh.
For now I'll close this now that everything's completed! ^_^

...OH I like to ask...
do you have a favorite eroge VN? Also do you like loli?
That's all and see ya!

**Comment by Cosetto** [comment]
@Linden10 @satan53x "!" and "," work fine for me.
![image](https://github.com/user-attachments/assets/2e9df88b-650e-4875-9f97-18700256a728)

**Reply by satan53x** [reply]
Yeah, maybe the entire sentence needs to be surrounded by `"` if you wanna use those symbols.
The `"` I say is a string type, not the displayed `"`. Just like the csv text, need to use the `"` if you want a linebreak.
Not tested, just guess.
Because the quote are checked during parsing.
![image](https://github.com/user-attachments/assets/4a219d19-9abc-40b6-a01a-60f35208cf65)

**Reply by Cosetto** [reply]
Seems like we have to decompile/compile SEEN.TXT with [rldev](https://github.com/eglaysher/rldev) to use "!" and ","
![image](https://github.com/user-attachments/assets/3a92a87b-83f9-4a5e-9d2d-904090b01c57)

**Reply by satan53x** [reply]
Not sure what to show, can you try to use SE and add `\"` into json? 
First find a sentence that does not display `!`, then add two `\"` to the sentence in json. (It is best if there is no linebreak in sentence)

**Reply by Cosetto** [reply]
> Not sure what to show, can you try to use SE and add `\"` into json? First find a sentence that does not display `!`, then add two `\"` to the sentence in json.

You're right, "!" and "," only display in ""
![image](https://github.com/user-attachments/assets/e8875e42-9422-467a-9ec2-a3a68d362e2b)

**Reply by Cosetto** [reply]
@satan53x Seen1001.txt by SExtractor and Rldev. Can you check to see the different?
![image](https://github.com/user-attachments/assets/d64045b0-a017-43b7-9cc2-d9b639f40dbd)
[Seen1001.zip](https://github.com/user-attachments/files/17784497/Seen1001.zip)
Update:
This is Seen1001.txt with narrative line has no ""
[Seen1001.zip](https://github.com/user-attachments/files/17784510/Seen1001.zip)

**Reply by satan53x** [reply]
This is normal.
SE only replace the text and fix the addr jump.
Rldev seems to automatically add invisible `"` to all text and re-compile the script. No need to add quotes manually.

**Reply by Cosetto** [reply]
> Rldev seems to automatically add invisible `"` to all text and re-compile the script. No need to add quotes manually.

Well damn, Rldev is totally useless against later Reallive version of Hadashi Shoujo 🤔
